### PR TITLE
feat: Signal watch

### PR DESCRIPTION
**Pull request by Yofou**
# Proposal
Signals currently can be render optimized so they don't rerun the component every time the value changes, we've extended this with `computed` allowing us to render optimize cases where you would usually use `useMemo`, however this introduce cases where can't utilize `useEffect` as that requires the component to rerun, hence we need to create a signal friendly equivalent.

# Usage
```tsx
const count = signal(0)
watch(() => {
  console.log('count trigger with ', count.value)
})
```

# Questions before further development
* Should there be a cleanup functionality like `useEffect`
> ANSWER: we decided that'd be the smart idea
* Should we trigger watch in a `queueEffect` when inside a component execution.
> ANSWER:
* Should we handle dep array changes (if the deps array changes.)
> ANSWER: The deps array was scratched as we couldn't come up with a compelling reason to add it at this time.

### CODE REVIEW

**Comment by LankyMoose** [comment]
@Yofou how attached are you to this deps array concept? I'd be happier if we could scrap that, and let users more explicitly opt out of implicit reactivity by way of ``signal.peek()``.

**Comment by Yofou** [comment]
> @Yofou how attached are you to this deps array concept? I'd be happier if we could scrap that, and let users more explicitly opt out of implicit reactivity by way of `signal.peek()`.

I'm not 100% attached, but it certainly comes in useful when you're doing conditional reactivity.

ie take a look at this simple case.
```tsx
const isTracking = signal(false)
const counter1 = signal(0)
const counter2 = signal(1)

watch(() => {
  if (isTracking.value) {
    console.log(counter1.value)
  } else {
    console.log(counter2.value)
  }
})
```

watch will only trigger on `isTracking` and `counter1` changes if `isTracking` is true and when `isTracking` is false, it will not longer track `counter1` but track `counter2` instead. which means the user may hit a pitfall where they'd get "missed" watch callbacks due to this, without having a dependencies array as an option the developer will have two options.

```tsx
watch(() => {
  // even though on sight this looks useless, this essentially ensure we always track these.
  counter2.value
  counter1.value

  if (isTracking.value) {
    console.log(counter1.value)
  } else {
    console.log(counter2.value)
  }
})
```

or option 2 which is handle the subscriptions to these signal themselves, with the same callback and ensure they handle cleanup themselves too.

but regardless even though I'd agree that a dependency array feature is not needed. it would be incredibly more convenient to allow them to explicitly set which signals they listen too in an more idiomatic manner and potentially save developers not cleaning up subscription properly.

**Comment by LankyMoose** [comment]
Ok, call me convinced - hadn't considered cases like that. Happy to go ahead with this API but would like to see the deps array function more like it does with hooks like ``useEffect`` when used during component execution 👌

**Comment by LankyMoose** [comment]
On second thought, I am no longer convinced. 

```ts
const isTracking = signal(false)
const counter1 = signal(0)
const counter2 = signal(1)

watch(() => {
  if (isTracking.value) {
    console.log(counter1.value)
  } else {
    console.log(counter2.value)
  }
})
```
Perhaps the watch deps *should* be fluid, ie why should we care about ``counter1`` in this scenario? Should changes to ``counter1`` cause the callback to fire? I don't think so 🤷‍♂️

**Comment by Yofou** [comment]
> On second thought, I am no longer convinced.
> 
> ```ts
> const isTracking = signal(false)
> const counter1 = signal(0)
> const counter2 = signal(1)
> 
> watch(() => {
>   if (isTracking.value) {
>     console.log(counter1.value)
>   } else {
>     console.log(counter2.value)
>   }
> })
> ```
> 
> Perhaps the watch deps _should_ be fluid, ie why should we care about `counter1` in this scenario? Should changes to `counter1` ever cause the callback to fire if it has not been observed? I don't think so 🤷‍♂️

@LankyMoose 

haha alright call me convinced then. I stand by my statement if the user does want this then they still have options I suppose but it may be rarer than I originally thought so we don't have to support it out right.

I've updated the Questions section in this thread with this conclusion and will pull out the deps array logic. soon.
how do you feel about these questions though

* Should there be a cleanup functionality like useEffect
* Should we trigger watch in a queueEffect when inside a component execution.

my generel feeling is we should have a cleanup function for component execution similar to useEffect.

but not sure about "Should we trigger watch in a queueEffect when inside a component execution."

**Comment by LankyMoose** [comment]
Definitely agree regarding cleanup, also not sure about whether it should fire during render or post-flush. Perhaps it's worth demonstrating some real world examples?

**Comment by Yofou** [comment]
> Definitely agree regarding cleanup, also not sure about whether it should fire during render or post-flush. Perhaps it's worth demonstrating some real world examples?

I'll play around with the queueEffect idea, my guess is it'd be for the primary suspect of using refs inside watch, but I'm not sure how queueEffect behaves when calling during a non component execution happening, yk what I mean. but I'll guess I'll find out soon.

File: packages/lib/src/signal.ts [hunk 1]
```diff
@@ -127,14 +213,21 @@ export class Signal<T> {
           return this as Signal<any>
         },
         inject: (prev) => {
+          this.#subscribers
```
Comment (LankyMoose) [inline packages/lib/src/signal.ts:216]: 0.o

Comment (Yofou) [reply packages/lib/src/signal.ts]: Nice catch thanks!

**Review by LankyMoose — APPROVED** [review]
LGTM 😁🫎

**Pull request merged**